File: mergesort.py
def mergesort(arr: list[int])->list[int]:

    #If an array is not a single element, recursively split it into two arrays 
    if len(arr) > 1:
        leftArr = arr[:len(arr)//2]
        rightArr = arr[len(arr)//2:]
        
        #recursion
        mergesort(leftArr)
        mergesort(rightArr)

        #Now you merge the arrays
        i = 0 #left array index
        j = 0 #Right array index
        k = 0 #Index for merged array

        #Compare the left array and the right array and move the smaller one to the merged array
        while(i<len(leftArr) and j<len(rightArr)):
            if (leftArr[i]<rightArr[j]):
                arr[k] = leftArr[i]
                i +=1
            else:
                arr[k] = rightArr[j]
                j += 1
            k += 1
        
        #Once you've transferred all values from one of the arrays, add the remaining ones to the merged array
        while (i<len(leftArr)):
            arr[k] = leftArr[i]
            i += 1
            k += 1
        while(j<len(rightArr)):
            arr[k] =  rightArr[j]
            j += 1
            k += 1

             

    return arr

arr = [1,546,1,675,156425,12,52,-4,624,0,41,54]
print(mergesort(arr))

File: mergesort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> mergesort(vector<int>&nums){
    if(nums.size()>1){
        vector<int> left = std::vector<int>(nums.begin(),nums.end()-(nums.size()/2));
        vector<int> right;
        if(nums.size()%2==0){
            right = std::vector<int>(nums.begin()+(nums.size()/2), nums.end());
        }
        else{
            right = std::vector<int>(nums.begin()+(nums.size()/2)+1, nums.end());
        }

        // cout<<"--------LEFT---------"<<endl;
        // for(int i{0};i<left.size();i++){
        // cout<<"Element at "<<i<<" = "<<left.at(i)<<endl;
        // }

        // cout<<"--------RIGHT---------"<<endl;
        // for(int i{0};i<right.size();i++){
        // cout<<"Element at "<<i<<" = "<<right.at(i)<<endl;
        // }

        mergesort(left);
        mergesort(right);

        int i{0},j{0},k{0};

        while((i<left.size()) && (j<right.size())){
            if(left.at(i)<right.at(j)){
                nums.at(k) = left.at(i);
                i += 1;
            }
            else{
                nums.at(k) = right.at(j);
                j += 1;
            }
            k += 1;
        }

        while(i<left.size()){
            nums.at(k) = left.at(i);
            i += 1;
            k += 1;
        }

        while(j<right.size()){
            nums.at(k) = right.at(j);
            j += 1;
            k += 1;
        }
    }

    return nums;
}

// Driver code
int main()
{
    vector<int> arr = { 12, 11, 13, 5, 6, 7 };

    mergesort(arr);

    for(int i{0};i<arr.size();i++){
        cout<<"Element at "<<i<<" = "<<arr.at(i)<<endl;
    }
}